File: workspace/2023-03-12/factorial.rust/factorial.rs

fn main()
{
    for i in 0..20 {
        println!("factorial({}) = {}", i, factorial(i));
    }
}

fn factorial(n: i32) -> i32
{
    let mut a = 0;
    let mut b = 1;

    for _ in 0..n {
        b = a + b;
        a = b - a;
    }

    a
}

File: workspace/2025/2025-12-28/numerical-sequence.forth/factorial.fs

: _factorial ( fm n -- fn )
    dup 0 = if
        drop    ( fm n -- fn=fm )
    else
        tuck    ( fm n -- n fm n )
        *       ( n fm n -- n fm*n )
        swap    ( n fm*n -- fm*n n )
        1 -     ( fm*n n-1 )
        recurse
    endif
;

: factorial ( n -- fn )
    1           ( n -- n fm )
    swap        ( n fm -- fm n )
    _factorial  ( fm n -- fn )
;

: print_factorials_range ( max min -- )
    over over = if
        drop drop   ( max min -- )
    else
        dup         ( max min -- max min n )
        factorial . ( max min n -- max min )
        1 +         ( max min -- max min+1 )
        recurse
    endif
;

: print_factorials ( n -- )
    s\" factorials:\n" type
    1                       ( n -- max=n min )
    print_factorials_range  ( max=n min -- )
    s\" \n" type
;

